spirv_glsl.cpp

```diff
@@ -3348,100 +3305,285 @@ void CompilerGLSL::emit_extension_workarounds()
 
-		statement("uint subgroupBallotFindLSB(uvec4 value) { return findLSB(value.x); }");
-		statement("uint subgroupBallotFindMSB(uvec4 value) { return findMSB(value.x); }");
+			for (uint32_t i = 0u; i < exts.size(); ++i)
+			{
+				const Supp::Candidate e = exts[i];
+				const std::string name = Supp::getExtensionName(e);
 
-		statement("#else");
+				statement(i == 0u ? "#if" : "#elif", " defined(", name, ")");
+				switch (e)
+				{
+				case Supp::NV_shader_thread_group:
+					statement("#define gl_SubgroupID gl_WarpIDNV");
+					break;
+				}
+			}
+			statement("#endif");
+		}
+		else if (shader_subgroup_supporter.isFeatureRequested(Supp::NumSubgroups))
+		{
+			auto exts = shader_subgroup_supporter.getCandidatesForFeature(Supp::NumSubgroups, result);
 
-		statement("uint subgroupBallotFindLSB(uvec4 value) {");
-		statement("int firstLive = findLSB(value.x);");
-		statement("return uint(firstLive != -1 ? firstLive : (findLSB(value.y) + 32));");
-		statement("}");
-		statement("uint subgroupBallotFindMSB(uvec4 value) {");
-		statement("int firstLive = findMSB(value.y);");
-		statement("return uint(firstLive != -1 ? (firstLive + 32) : findMSB(value.x));");
-		statement("}");
+			for (uint32_t i = 0u; i < exts.size(); ++i)
+			{
+				const Supp::Candidate e = exts[i];
+				const std::string name = Supp::getExtensionName(e);
 
-		statement("#endif");
-	}
-	// GL_KHR_shader_subgroup_vote
-	if (!options.vulkan_semantics && declared_KHR_shader_subgroup_vote)
-	{
-		statement("#ifdef GL_ARB_shader_group_vote");
-		statement("bool subgroupAll(bool v) { return allInvocationsARB(v); }");
-		statement("bool subgroupAny(bool v) { return anyInvocationARB(v); }");
-		statement("bool subgroupAllEqual(bool v) { return allInvocationsEqualARB(v); }");
-		statement("#define SUBGROUP_ALL_EQUAL_WORKAROUND(type) bool subgroupAllEqual(type value) { return "
-		          "subgroupAllEqual(subgroupBroadcastFirst(value)==value); }");
-		for (const char *t : TYPES)
+				statement(i == 0u ? "#if" : "#elif", " defined(", name, ")");
+				switch (e)
+				{
+				case Supp::NV_shader_thread_group:
+					statement("#define gl_NumSubgroups gl_WarpsPerSMNV");
+					break;
+				}
+			}
+			statement("#endif");
+		}
+		else if (shader_subgroup_supporter.isFeatureRequested(Supp::SubgroupBrodcast_First))
 		{
-			statement(std::string("SUBGROUP_ALL_EQUAL_WORKAROUND(") + t + ")");
+			auto exts = shader_subgroup_supporter.getCandidatesForFeature(Supp::SubgroupBrodcast_First, result);
+
+			for (uint32_t i = 0u; i < exts.size(); ++i)
+			{
+				const Supp::Candidate e = exts[i];
+				const std::string name = Supp::getExtensionName(e);
+
+				statement(i == 0u ? "#if" : "#elif", " defined(", name, ")");
+				switch (e)
+				{
+				case Supp::NV_shader_thread_shuffle:
+					for (const char *t : TYPES)
+						statement(string(t) + " subgroupBroadcastFirst(" + t +
+						          " value) { return shuffleNV(value, 0, gl_WarpSizeNV); }");
```

Comment:
`shuffleNV` has slightly different semantics to `subgroupBroadcastFirst` and `readFirstInvocationARB`, you'd need to use ` findLSB(ballotThreadNV(true))` instead of 0 here.